Adding multicolumn cupertino picker to the gallery.

doxuanviet1996 opened the pull request.

Commit:
adding multicolumn_picker to gallery

Minor fix in documentation

Add flexes to multicolumn_picker

Fix the magnifier

Clean up code + comments

Minor fix

minor fix

Mark flutter_gallery_instrumentation_test as flaky (#18887)

See #18879.

Fix library-scope expression evaluation. (#18851)

This also improves diagnostic in case of internal compiler errors.

Remove race conditions involving finding available ports (#18698)

This is an attempt to reland #18488 with less breakage on macOS.

Filter invalid characters in the terminal input stream (#18892)

Fixes https://github.com/flutter/flutter/issues/18007

Add --create option to `flutter emulators` command (#18235)

* Add --create option to flutter emulators

* Tweaks to error message

* Simplify emulator search logic

* Make name optional

* Add a note about this option being used with --create

* Tweaks to help information

* Switch to processManager for easier testing

* Don't crash on missing files or missing properties in Android Emulator

* Move name suffixing into emulator manager

This allows it to be tested in the EmulatorManager tests and also used by daemon later if desired.

* Pass the context's android SDK through so it can be mocked by tests

* Misc fixes

* Add tests around emulator creation

Process calls are mocked to avoid needing a real SDK (and to be fast). Full integration tests may be useful, but may require ensuring all build environments etc. are set up correctly.

* Simplify avdManagerPath

Previous changes were to emulatorPath!

* Fix lint errors

* Fix incorrect file exgtension for Windows

* Fix an issue where no system images would crash

reduce throws on an empty collection.

* Fix "null" appearing in error messages

The name we attempted to use will now always be returned, even in the case of failure.

* Add additional info to missing-system-image failure message

On Windows after installing Andriod Studio I didn't have any of these and got this message. Installing with sdkmanager fixed the issue.

* Fix thrown errors

runResult had a toString() but we moved to ProcessResult when switching to ProcessManager to this ended up throwing "Instance of ProcessResult".

* Fix package import

* Fix more package imports

* Move mock implementation into Mock class

There seemed to be issues using Lists in args with Mockito that I couldn't figure out (docs say to use typed() but I couldn't make this compile with these lists still)..

* Rename method that's ambigious now we have create

* Handle where there's no avd path

* Add another toList() :(

* Remove comment that was rewritten

* Fix forbidden import

* Make optional arg more obviously optional

* Reformat doc

* Note that we create a pixel device in help text

* Make this a named arg

Add process output for flutter_tester test and unskip (#18899)

* Add process output for flutter_tester test and unskip

* Add a comment about this potential flake

Add iOS module template (#18830)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Re-skip flutter_device_tester test (#18900)

After landing the un-skip this test failed with a timeout. It then passed on the next build (!). I think it's too flaky to leave running until we can better diagnose what's happening.

Revert "Add iOS module template (#18830)" (#18902)

This reverts commit 8be198d8a086bc26a9b73536f5ebd7ee327a4f01 that broke the build.

Ensure we have the Java binary on PATH when invoking avdmanager (#18908)

See https://github.com/flutter/flutter/issues/13379#issuecomment-400984667.

Add iOS module template (#18830) (#18903)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Wait before running the second test. (#18826)

Expose emulator creation to daemon API (#18905)

* Fix typo

* Add emulator.create to Daemon API

* Swap order of daemon changelog entries

Remove the 'app' domain from flutter daemon (#18873)

* Remove the 'app' domain from flutter daemon

By default the daemon won't register the "app" domain, you need to opt-in (which the 'run' command does, as well as the tests for the app functionality).

Fixes #6658.

* Tweak text

* Put restart/callServiceExtension/stop back into daemon mode

* Add a comment about removing discoverApps

Make Flutter module template use FlutterView with less assumptions (#18883)

Remove flaky flutter_tester test (#18916)

This test was designed to ensure flutter_tester keeps running (previously it would quit immediately). However it's turned out ot be rather flaky and we have new tests on the way that supersede this by actually testing real things (debug stepping, reloading, expression evaluation).

Provide workaround for expression compilation frontend issue (#18657)

Add 90th and 99th percentile frame rasterize time for timeline_summary (#18881)

Our first step towards https://github.com/flutter/flutter/issues/18727

Add golden test for text shade overflow (#18816)

Goldens for flutter/flutter#18729

Add 90th and 99th percentiles to dashboard (#18930)

This fixes #18727

Fix typo (#18941)

Set FLUTTER_TEST when spawning flutter_tester (#18907)

Otherwise on Windows these will fail as an unsupported platform when things try to read defaultTargetPlatform in flutter apps.

Retry npm step 5 times with 5 second pause (#18918)

Mark non-flaky tests as such (#18927)

Let the Kernel FE know whether it's building for a release VM. (#18876)

Fixes #18829

Fix intrinsics for ConstrainedBox (#18935)

Fix type errors in iOS simulator discovery. (#18960)

Support all keyboard actions. (#11344) (#18855)

* Support all keyboard actions. (#11344)

Simplify Gradle magic (#18987)

Fix module_test (#18993)

Throw instead of return null if durations is empty (#18957)

See https://github.com/flutter/flutter/pull/18881#discussion_r199285256

[frdb] Change to warning log on timeout. (#19006)

Before a severe log would be raised when timeout happens. Now that
testing connections requires potentially running into a timeout, this
will cause failures when there shouldn't be any.

Show Search Button on keyboard for search widget (#19011)

Also:
 * Export TextInputAction from `text_field.dart` to parallel `TextInputType`.

Nit: add space after for (#18954)

Round properly to nearest page in TabBarView. (#16868)

This logic is described in the test as looking for a scroll ending
very close to a new page, but in fact its behavior is more like
"very close to a page to the right": if we're not very, very close
to any page, it will pick the page to the left, not an old page.

There's no reason this should be left-right asymmetrical.
Instead, pick the nearest page.

In practice, the case where this makes a difference never arises when
the scroll runs undisturbed to completion; but when the user taps on
the page to hold or drag, the scroll will be interrupted before it
gets within tolerance of a particular page, and this case does arise.

This fixes a glitch that is hard to trigger without time dilation,
but is quite conspicuous with it:

 * Open a tab view with at least 4 tabs, e.g. the Buttons screen
   of the gallery (with "Animate Slowly" on.)
 * Starting at tab 0, tap tab 2.
 * When the animation is nearly complete, tap the page a couple
   of times, as if to drag it around to scroll.  Then let the
   page view settle ballistically toward page 2.
 * Before it finishes, tap tab 3.
 * Suddenly page 1 fills the view, replacing page 2, before we
   scroll from there to page 3.

With this fix, the animation in the last step moves smoothly from
where we are when it starts onward to page 3.

Fix type for Dart 2 (#18998)

Fixes #18995.

Include stdout/stderr in failure messages for devicelab's tests (#18990)

* Include stdout/stderr in failure messages

Sometimes some of these tests unexpectedly fail with a non-zero exit code. This ensures stdout/stderr is included in the test failure message when this happens so that we can track down the issue.

* Remove redundant info about exit code

* Remove unnecessary indenting

The [markers] make it fairly clear so this just makes the test code noisy.

flutter attach should handle non-singular device gracefully (#19036)

Upgrade node on travis (#18994)

* Upgrade to current LTS version of Node

The version of NPM with Node 6 is over a year old and frequently hangs trying to install Firebase tools.

* Temporarily comment out this condition so the install runs on Travis

* Revert "Temporarily comment out this condition so the install runs on Travis"

This reverts commit 74db9366b4b1b094369388ef50fb26f3b7fd3104.

* Pin to nodejs v8.1

Add flag to ThemeData to expand tap targets of certain material widgets (#18369)

Make EditableText cursor configurable (#18888)

* fixed segmented control golden test

* fixed segmented control golden test

* added cursorWidth, cursorRadius

* added default value for cursorWidth based on Apple specs

* test default cursorWidth

* removed cursorHeight stuff

* added functionality to keep cursor from blinking

* cursor width and radius is configurable + tests

* changed goldens repo version in goldens.version

* working version of configurable cursor (erased debugKeepCursorOn)

* minor changes

* docs

* changed textfield test that was failing due to new default cursorwidth

* added default value of cursorwidth in RenderEditable

* only run golden file tests on Mac

* cursor tests

* the tests are actually there now

* weak warning fixed

* switching to Linux

* changed default cursorWidth: 2.0 -> 1.0

* assorted changes, including changing text field test

* re-paint -> re-layout when changing cursorWidth

Revert "Add flag to ThemeData to expand tap targets of certain material widgets (#18369)" (#19053)

This reverts commit 989f5741da9e4e8a9ae40d2ec9fbe33f11aba177.

make edit in README.md (#18981)

Fix typos (#19065)

Fix typos (#19067)

Fix typos (#19073)

Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd (#19008)

* Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd

flutter/engine@21c7d6a Revert "Add antiAlias and saveCount to clipPath and restore (#5638)"
flutter/engine@70dcbb5 Roll src/third_party/skia d818ebf4a317..9aa30c6ee0e5 (11 commits)
flutter/engine@ad42324 Roll src/third_party/skia a219419c9d76..d818ebf4a317 (2 commits)
flutter/engine@f2eb83a Roll src/third_party/skia 00d2e8ebcb13..a219419c9d76 (1 commits)
flutter/engine@0ec7669 Roll src/third_party/skia 8451daabb23d..00d2e8ebcb13 (1 commits)
flutter/engine@061e899 Support all keyboard actions. (#11344)
flutter/engine@bc6b250 Roll src/third_party/skia 34024a7c478c..8451daabb23d (1 commits)
flutter/engine@df4dffb Roll src/third_party/skia 75e69028956d..34024a7c478c (1 commits)
flutter/engine@d8770d4 Roll src/third_party/skia cf863fb9b446..75e69028956d (1 commits)
flutter/engine@b88a8b3 Roll src/third_party/skia b7b9d02ac020..cf863fb9b446 (1 commits)
flutter/engine@2261ccf [fuchsia] Update scenic lib path.
flutter/engine@4c4ef98 [fuchsia] Rename scenic_lib => scenic
flutter/engine@f0c21f3 Roll src/third_party/skia 184d408b646b..b7b9d02ac020 (7 commits)
flutter/engine@a2bf805 Add antiAlias and saveCount to clipPath and restore
flutter/engine@9e450d1 Roll src/third_party/skia c91fe3ab1c5d..184d408b646b (10 commits)
flutter/engine@e663996 Roll src/third_party/skia eb8f8106f38c..c91fe3ab1c5d (5 commits)
flutter/engine@fecd66f Roll src/third_party/skia 723b1f6ef941..eb8f8106f38c (1 commits)
flutter/engine@4466d61 Remove vmservice_io.main from entry points.
flutter/engine@f279dfe Roll src/third_party/skia 7e2327b133db..723b1f6ef941 (1 commits)
flutter/engine@a885bd4 Roll src/third_party/skia 24d18ced1ad7..7e2327b133db (6 commits)
flutter/engine@ad1bd47 Roll src/third_party/skia a1e5630183c1..24d18ced1ad7 (7 commits)
flutter/engine@ce06bba ensure a11y state is communicated back to flutter
flutter/engine@c54a57e Roll src/third_party/skia b1b87d9df81e..a1e5630183c1 (10 commits)
flutter/engine@a9e5354 Roll src/third_party/skia 385804514edf..b1b87d9df81e (6 commits)
flutter/engine@2006e7d Roll src/third_party/skia e97bb26893a1..385804514edf (1 commits)
flutter/engine@2942061 Roll src/third_party/skia e2e52e46ca63..e97bb26893a1 (1 commits)
flutter/engine@e378811 Roll src/third_party/skia 14de25dfc7c4..e2e52e46ca63 (2 commits)
flutter/engine@48cc8f2 Roll src/third_party/skia 551dc3e91143..14de25dfc7c4 (2 commits)
flutter/engine@00ddf13 libtxt: apply an offset to drawing operations instead of translating the canvas
flutter/engine@2c5647c Roll src/third_party/skia 059a9ab4bcd0..551dc3e91143 (5 commits)
flutter/engine@9e44004 Roll src/third_party/skia 45c9dab4c3ec..059a9ab4bcd0 (8 commits)
flutter/engine@fbb3436 Remove some unused code from the Android host
flutter/engine@36f3f95 Roll src/third_party/skia 37b7e4714558..45c9dab4c3ec (1 commits)
flutter/engine@57e2e0c Roll src/third_party/skia c421ca1d6e41..37b7e4714558 (2 commits)
flutter/engine@ce417e2 Roll src/third_party/skia 5b201e3b0ba8..c421ca1d6e41 (3 commits)
flutter/engine@05f5a18 Roll src/third_party/skia d47fe095cf88..5b201e3b0ba8 (6 commits)
flutter/engine@33b70e0 Roll src/third_party/skia 9c0ce41cf711..d47fe095cf88 (12 commits)
flutter/engine@7dab419 Roll src/third_party/skia d4b2adeaa929..9c0ce41cf711 (7 commits)
flutter/engine@d3fa01a Roll src/third_party/skia f46710802ad5..d4b2adeaa929 (1 commits)

Create a CupertinoApp (#18469)

Avoid repeating microbenchmark (#19087)

Attempt to fix race in microbenchmark error handling code (#19092)

Await stream cancellations before trying to complete

Add temporary debug prints to identify devicelab failures (#19093)

Revert "Add temporary debug prints to identify devicelab failures (#19093)" (#19095)

This reverts commit edc16473bfdafc0f5fe546610d3b0fe0eecdd080.

Revert engine roll that caused microbenchmarks failures (#19123)

* Merge

* Update goldens

Adding a setup function for hot restart (#18921)

* Adding a setup function for hot restart that can be overwritten to perform any necessary initialization before a restart is started.

* Adding a test for setupHotRestart

* Removing local pubspec change

* Fixing some type annotation errors

* Responding to comments

Deferred ListView.separator cleanups (#18720)

Fix typo (#19160)

Additional integration test driver functionality (#19085)

* Add additional functionality to integration test driver

* Typo

* Remove delays we shouldn't need + comments

* Fix typos

Fix path of ios module template files (#19037)

* Fix path of ios module template files

* Fix more paths

make tests in xcodeproj_test.dart and gradle_test.dart hermetic (#18093) (#18548)

Re-land #18369 (#19082)

adjust declared types to work with dart 2 typing at runtime (#19007)

* adjust declared types to work with dart 2 typing at runtime

* review comments

* update packages/flutter_tools/lib/src/ios/simulators.dart

remove an unused analysis options file (#19010)

Adjust Cupertino controls's APIs to match Material (#19127)

Put the nav bar padding inside the CustomMultiChildLayout for correct centering and allow custom padding (#19129)

Refer to Navigator.of(rootNavigator) in Cupertino tab widgets (#19130)

Move collectCoverageData into CoverageCollector so it can be re-used. (#19055)

This will let us re-use that code from fuchsia_tester.dart.
Tested by running the stocks example tests with coverage collection before and after.
.lcov files are identical.

Add Windows support to defaultTargetPlatform (#19176)

A significant amonut of code uses defaultTargetPlatform, so it's
currently impossible to run most apps on Windows. This adds a mapping
from Windows as a host to Android as a target, paralleling the
macOS->iOS and Linux->Android mappings.

This allows more use of Windows as a host platform (e.g., for testing,
as in issue #17768).

Add new matcher and utility methods for testing semanics (#19046)

Add getSemanticsId command to flutter_driver (#19047)

Use runTests in fuchsia tester. (#19178)

This allows us to collect code coverage for Fuchsia tests.

Rebuilding BottomNavigationBar when items.length changes (#19179)

Updated the material localizations (#19174)

Add ignoringFeedbackSemantics to Draggable/LongPressDraggable (#18643)

Remove bogus SDK constraint lines. (#19013)

Segmented control animation (#18811)

Added animation to segmented control widget.

Improve flutter tools integration tests (#18865)

* Rename util -> test_utils

* Rename flutter_test_driver -> test_driver

* Switch testWithContext -> test

* Remove unused import

* Move test project into a class to make it easier to have multiple of these

Each "TestProject" class can contain its files and things like named breakpoint locations.

* Split expression evaluation tests into own file

* Include last response in error messages

* Update expectations based on current bugs

* Fix async-ness in tests

* Fix incorrect expectation in test

* Fix incorrect evaluations

* Remove skips for tests that are now passing on master

* Expect pass on Linux

🤷‍♂️

* Call the code

* Skip expression evaluation tests on Windows

* Skip whole group, not just one test

* Remove duplicated method from merge

* Fix misplaced close of group

* Remove code that was duplicated from test we copied

Not sure how this ended up in here?

* Re-fix typo

Improve error message in microbenchmarks when flutter run lives too long (#19200)

This should make it more obvious from the error what's gone wrong if we see a repeat of #19096 (previously the error was something like "Bad state: future already completed").

Fix use of Future<void> as a Future<bool> (#18843)

* Fix use of Future<void> as a Future<bool>

Add segmented control to gallery (#19192)

* Adding segmented control to gallery

Commit:
adding multicolumn_picker to gallery

Minor fix in documentation

Add flexes to multicolumn_picker

Fix the magnifier

Clean up code + comments

Minor fix

minor fix

Mark flutter_gallery_instrumentation_test as flaky (#18887)

See #18879.

Fix library-scope expression evaluation. (#18851)

This also improves diagnostic in case of internal compiler errors.

Remove race conditions involving finding available ports (#18698)

This is an attempt to reland #18488 with less breakage on macOS.

Filter invalid characters in the terminal input stream (#18892)

Fixes https://github.com/flutter/flutter/issues/18007

Add --create option to `flutter emulators` command (#18235)

* Add --create option to flutter emulators

* Tweaks to error message

* Simplify emulator search logic

* Make name optional

* Add a note about this option being used with --create

* Tweaks to help information

* Switch to processManager for easier testing

* Don't crash on missing files or missing properties in Android Emulator

* Move name suffixing into emulator manager

This allows it to be tested in the EmulatorManager tests and also used by daemon later if desired.

* Pass the context's android SDK through so it can be mocked by tests

* Misc fixes

* Add tests around emulator creation

Process calls are mocked to avoid needing a real SDK (and to be fast). Full integration tests may be useful, but may require ensuring all build environments etc. are set up correctly.

* Simplify avdManagerPath

Previous changes were to emulatorPath!

* Fix lint errors

* Fix incorrect file exgtension for Windows

* Fix an issue where no system images would crash

reduce throws on an empty collection.

* Fix "null" appearing in error messages

The name we attempted to use will now always be returned, even in the case of failure.

* Add additional info to missing-system-image failure message

On Windows after installing Andriod Studio I didn't have any of these and got this message. Installing with sdkmanager fixed the issue.

* Fix thrown errors

runResult had a toString() but we moved to ProcessResult when switching to ProcessManager to this ended up throwing "Instance of ProcessResult".

* Fix package import

* Fix more package imports

* Move mock implementation into Mock class

There seemed to be issues using Lists in args with Mockito that I couldn't figure out (docs say to use typed() but I couldn't make this compile with these lists still)..

* Rename method that's ambigious now we have create

* Handle where there's no avd path

* Add another toList() :(

* Remove comment that was rewritten

* Fix forbidden import

* Make optional arg more obviously optional

* Reformat doc

* Note that we create a pixel device in help text

* Make this a named arg

Add process output for flutter_tester test and unskip (#18899)

* Add process output for flutter_tester test and unskip

* Add a comment about this potential flake

Add iOS module template (#18830)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Re-skip flutter_device_tester test (#18900)

After landing the un-skip this test failed with a timeout. It then passed on the next build (!). I think it's too flaky to leave running until we can better diagnose what's happening.

Revert "Add iOS module template (#18830)" (#18902)

This reverts commit 8be198d8a086bc26a9b73536f5ebd7ee327a4f01 that broke the build.

Ensure we have the Java binary on PATH when invoking avdmanager (#18908)

See https://github.com/flutter/flutter/issues/13379#issuecomment-400984667.

Add iOS module template (#18830) (#18903)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Wait before running the second test. (#18826)

Expose emulator creation to daemon API (#18905)

* Fix typo

* Add emulator.create to Daemon API

* Swap order of daemon changelog entries

Remove the 'app' domain from flutter daemon (#18873)

* Remove the 'app' domain from flutter daemon

By default the daemon won't register the "app" domain, you need to opt-in (which the 'run' command does, as well as the tests for the app functionality).

Fixes #6658.

* Tweak text

* Put restart/callServiceExtension/stop back into daemon mode

* Add a comment about removing discoverApps

Make Flutter module template use FlutterView with less assumptions (#18883)

Remove flaky flutter_tester test (#18916)

This test was designed to ensure flutter_tester keeps running (previously it would quit immediately). However it's turned out ot be rather flaky and we have new tests on the way that supersede this by actually testing real things (debug stepping, reloading, expression evaluation).

Provide workaround for expression compilation frontend issue (#18657)

Add 90th and 99th percentile frame rasterize time for timeline_summary (#18881)

Our first step towards https://github.com/flutter/flutter/issues/18727

Add golden test for text shade overflow (#18816)

Goldens for flutter/flutter#18729

Add 90th and 99th percentiles to dashboard (#18930)

This fixes #18727

Fix typo (#18941)

Set FLUTTER_TEST when spawning flutter_tester (#18907)

Otherwise on Windows these will fail as an unsupported platform when things try to read defaultTargetPlatform in flutter apps.

Retry npm step 5 times with 5 second pause (#18918)

Mark non-flaky tests as such (#18927)

Let the Kernel FE know whether it's building for a release VM. (#18876)

Fixes #18829

Fix intrinsics for ConstrainedBox (#18935)

Fix type errors in iOS simulator discovery. (#18960)

Support all keyboard actions. (#11344) (#18855)

* Support all keyboard actions. (#11344)

Simplify Gradle magic (#18987)

Fix module_test (#18993)

Throw instead of return null if durations is empty (#18957)

See https://github.com/flutter/flutter/pull/18881#discussion_r199285256

[frdb] Change to warning log on timeout. (#19006)

Before a severe log would be raised when timeout happens. Now that
testing connections requires potentially running into a timeout, this
will cause failures when there shouldn't be any.

Show Search Button on keyboard for search widget (#19011)

Also:
 * Export TextInputAction from `text_field.dart` to parallel `TextInputType`.

Nit: add space after for (#18954)

Round properly to nearest page in TabBarView. (#16868)

This logic is described in the test as looking for a scroll ending
very close to a new page, but in fact its behavior is more like
"very close to a page to the right": if we're not very, very close
to any page, it will pick the page to the left, not an old page.

There's no reason this should be left-right asymmetrical.
Instead, pick the nearest page.

In practice, the case where this makes a difference never arises when
the scroll runs undisturbed to completion; but when the user taps on
the page to hold or drag, the scroll will be interrupted before it
gets within tolerance of a particular page, and this case does arise.

This fixes a glitch that is hard to trigger without time dilation,
but is quite conspicuous with it:

 * Open a tab view with at least 4 tabs, e.g. the Buttons screen
   of the gallery (with "Animate Slowly" on.)
 * Starting at tab 0, tap tab 2.
 * When the animation is nearly complete, tap the page a couple
   of times, as if to drag it around to scroll.  Then let the
   page view settle ballistically toward page 2.
 * Before it finishes, tap tab 3.
 * Suddenly page 1 fills the view, replacing page 2, before we
   scroll from there to page 3.

With this fix, the animation in the last step moves smoothly from
where we are when it starts onward to page 3.

Fix type for Dart 2 (#18998)

Fixes #18995.

Include stdout/stderr in failure messages for devicelab's tests (#18990)

* Include stdout/stderr in failure messages

Sometimes some of these tests unexpectedly fail with a non-zero exit code. This ensures stdout/stderr is included in the test failure message when this happens so that we can track down the issue.

* Remove redundant info about exit code

* Remove unnecessary indenting

The [markers] make it fairly clear so this just makes the test code noisy.

flutter attach should handle non-singular device gracefully (#19036)

Upgrade node on travis (#18994)

* Upgrade to current LTS version of Node

The version of NPM with Node 6 is over a year old and frequently hangs trying to install Firebase tools.

* Temporarily comment out this condition so the install runs on Travis

* Revert "Temporarily comment out this condition so the install runs on Travis"

This reverts commit 74db9366b4b1b094369388ef50fb26f3b7fd3104.

* Pin to nodejs v8.1

Add flag to ThemeData to expand tap targets of certain material widgets (#18369)

Make EditableText cursor configurable (#18888)

* fixed segmented control golden test

* fixed segmented control golden test

* added cursorWidth, cursorRadius

* added default value for cursorWidth based on Apple specs

* test default cursorWidth

* removed cursorHeight stuff

* added functionality to keep cursor from blinking

* cursor width and radius is configurable + tests

* changed goldens repo version in goldens.version

* working version of configurable cursor (erased debugKeepCursorOn)

* minor changes

* docs

* changed textfield test that was failing due to new default cursorwidth

* added default value of cursorwidth in RenderEditable

* only run golden file tests on Mac

* cursor tests

* the tests are actually there now

* weak warning fixed

* switching to Linux

* changed default cursorWidth: 2.0 -> 1.0

* assorted changes, including changing text field test

* re-paint -> re-layout when changing cursorWidth

Revert "Add flag to ThemeData to expand tap targets of certain material widgets (#18369)" (#19053)

This reverts commit 989f5741da9e4e8a9ae40d2ec9fbe33f11aba177.

make edit in README.md (#18981)

Fix typos (#19065)

Fix typos (#19067)

Fix typos (#19073)

Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd (#19008)

* Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd

flutter/engine@21c7d6a Revert "Add antiAlias and saveCount to clipPath and restore (#5638)"
flutter/engine@70dcbb5 Roll src/third_party/skia d818ebf4a317..9aa30c6ee0e5 (11 commits)
flutter/engine@ad42324 Roll src/third_party/skia a219419c9d76..d818ebf4a317 (2 commits)
flutter/engine@f2eb83a Roll src/third_party/skia 00d2e8ebcb13..a219419c9d76 (1 commits)
flutter/engine@0ec7669 Roll src/third_party/skia 8451daabb23d..00d2e8ebcb13 (1 commits)
flutter/engine@061e899 Support all keyboard actions. (#11344)
flutter/engine@bc6b250 Roll src/third_party/skia 34024a7c478c..8451daabb23d (1 commits)
flutter/engine@df4dffb Roll src/third_party/skia 75e69028956d..34024a7c478c (1 commits)
flutter/engine@d8770d4 Roll src/third_party/skia cf863fb9b446..75e69028956d (1 commits)
flutter/engine@b88a8b3 Roll src/third_party/skia b7b9d02ac020..cf863fb9b446 (1 commits)
flutter/engine@2261ccf [fuchsia] Update scenic lib path.
flutter/engine@4c4ef98 [fuchsia] Rename scenic_lib => scenic
flutter/engine@f0c21f3 Roll src/third_party/skia 184d408b646b..b7b9d02ac020 (7 commits)
flutter/engine@a2bf805 Add antiAlias and saveCount to clipPath and restore
flutter/engine@9e450d1 Roll src/third_party/skia c91fe3ab1c5d..184d408b646b (10 commits)
flutter/engine@e663996 Roll src/third_party/skia eb8f8106f38c..c91fe3ab1c5d (5 commits)
flutter/engine@fecd66f Roll src/third_party/skia 723b1f6ef941..eb8f8106f38c (1 commits)
flutter/engine@4466d61 Remove vmservice_io.main from entry points.
flutter/engine@f279dfe Roll src/third_party/skia 7e2327b133db..723b1f6ef941 (1 commits)
flutter/engine@a885bd4 Roll src/third_party/skia 24d18ced1ad7..7e2327b133db (6 commits)
flutter/engine@ad1bd47 Roll src/third_party/skia a1e5630183c1..24d18ced1ad7 (7 commits)
flutter/engine@ce06bba ensure a11y state is communicated back to flutter
flutter/engine@c54a57e Roll src/third_party/skia b1b87d9df81e..a1e5630183c1 (10 commits)
flutter/engine@a9e5354 Roll src/third_party/skia 385804514edf..b1b87d9df81e (6 commits)
flutter/engine@2006e7d Roll src/third_party/skia e97bb26893a1..385804514edf (1 commits)
flutter/engine@2942061 Roll src/third_party/skia e2e52e46ca63..e97bb26893a1 (1 commits)
flutter/engine@e378811 Roll src/third_party/skia 14de25dfc7c4..e2e52e46ca63 (2 commits)
flutter/engine@48cc8f2 Roll src/third_party/skia 551dc3e91143..14de25dfc7c4 (2 commits)
flutter/engine@00ddf13 libtxt: apply an offset to drawing operations instead of translating the canvas
flutter/engine@2c5647c Roll src/third_party/skia 059a9ab4bcd0..551dc3e91143 (5 commits)
flutter/engine@9e44004 Roll src/third_party/skia 45c9dab4c3ec..059a9ab4bcd0 (8 commits)
flutter/engine@fbb3436 Remove some unused code from the Android host
flutter/engine@36f3f95 Roll src/third_party/skia 37b7e4714558..45c9dab4c3ec (1 commits)
flutter/engine@57e2e0c Roll src/third_party/skia c421ca1d6e41..37b7e4714558 (2 commits)
flutter/engine@ce417e2 Roll src/third_party/skia 5b201e3b0ba8..c421ca1d6e41 (3 commits)
flutter/engine@05f5a18 Roll src/third_party/skia d47fe095cf88..5b201e3b0ba8 (6 commits)
flutter/engine@33b70e0 Roll src/third_party/skia 9c0ce41cf711..d47fe095cf88 (12 commits)
flutter/engine@7dab419 Roll src/third_party/skia d4b2adeaa929..9c0ce41cf711 (7 commits)
flutter/engine@d3fa01a Roll src/third_party/skia f46710802ad5..d4b2adeaa929 (1 commits)

Create a CupertinoApp (#18469)

Avoid repeating microbenchmark (#19087)

Attempt to fix race in microbenchmark error handling code (#19092)

Await stream cancellations before trying to complete

Add temporary debug prints to identify devicelab failures (#19093)

Revert "Add temporary debug prints to identify devicelab failures (#19093)" (#19095)

This reverts commit edc16473bfdafc0f5fe546610d3b0fe0eecdd080.

Revert engine roll that caused microbenchmarks failures (#19123)

* Merge

* Update goldens

Adding a setup function for hot restart (#18921)

* Adding a setup function for hot restart that can be overwritten to perform any necessary initialization before a restart is started.

* Adding a test for setupHotRestart

* Removing local pubspec change

* Fixing some type annotation errors

* Responding to comments

Deferred ListView.separator cleanups (#18720)

Fix typo (#19160)

Additional integration test driver functionality (#19085)

* Add additional functionality to integration test driver

* Typo

* Remove delays we shouldn't need + comments

* Fix typos

Fix path of ios module template files (#19037)

* Fix path of ios module template files

* Fix more paths

make tests in xcodeproj_test.dart and gradle_test.dart hermetic (#18093) (#18548)

Re-land #18369 (#19082)

adjust declared types to work with dart 2 typing at runtime (#19007)

* adjust declared types to work with dart 2 typing at runtime

* review comments

* update packages/flutter_tools/lib/src/ios/simulators.dart

remove an unused analysis options file (#19010)

Adjust Cupertino controls's APIs to match Material (#19127)

Put the nav bar padding inside the CustomMultiChildLayout for correct centering and allow custom padding (#19129)

Refer to Navigator.of(rootNavigator) in Cupertino tab widgets (#19130)

Move collectCoverageData into CoverageCollector so it can be re-used. (#19055)

This will let us re-use that code from fuchsia_tester.dart.
Tested by running the stocks example tests with coverage collection before and after.
.lcov files are identical.

Add Windows support to defaultTargetPlatform (#19176)

A significant amonut of code uses defaultTargetPlatform, so it's
currently impossible to run most apps on Windows. This adds a mapping
from Windows as a host to Android as a target, paralleling the
macOS->iOS and Linux->Android mappings.

This allows more use of Windows as a host platform (e.g., for testing,
as in issue #17768).

Add new matcher and utility methods for testing semanics (#19046)

Add getSemanticsId command to flutter_driver (#19047)

Use runTests in fuchsia tester. (#19178)

This allows us to collect code coverage for Fuchsia tests.

Rebuilding BottomNavigationBar when items.length changes (#19179)

Updated the material localizations (#19174)

Add ignoringFeedbackSemantics to Draggable/LongPressDraggable (#18643)

Remove bogus SDK constraint lines. (#19013)

Segmented control animation (#18811)

Added animation to segmented control widget.

Improve flutter tools integration tests (#18865)

* Rename util -> test_utils

* Rename flutter_test_driver -> test_driver

* Switch testWithContext -> test

* Remove unused import

* Move test project into a class to make it easier to have multiple of these

Each "TestProject" class can contain its files and things like named breakpoint locations.

* Split expression evaluation tests into own file

* Include last response in error messages

* Update expectations based on current bugs

* Fix async-ness in tests

* Fix incorrect expectation in test

* Fix incorrect evaluations

* Remove skips for tests that are now passing on master

* Expect pass on Linux

🤷‍♂️

* Call the code

* Skip expression evaluation tests on Windows

* Skip whole group, not just one test

* Remove duplicated method from merge

* Fix misplaced close of group

* Remove code that was duplicated from test we copied

Not sure how this ended up in here?

* Re-fix typo

Improve error message in microbenchmarks when flutter run lives too long (#19200)

This should make it more obvious from the error what's gone wrong if we see a repeat of #19096 (previously the error was something like "Bad state: future already completed").

Fix use of Future<void> as a Future<bool> (#18843)

* Fix use of Future<void> as a Future<bool>

Add segmented control to gallery (#19192)

* Adding segmented control to gallery

xster commented:
meta: step 1 for #7377

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
     _hasScrolled();
   }
 
+  /// Visual effect based on how we see the wheel. Imagine we are at some
```
Reading this it's easy to understand the effect from a relative sense but it's hard to predict the effect from an absolute sense. Can we make it easier to quantify this effect (either by making equivalents in the physical world etc)? In other word, how can I mentally visualize what would happen if I put in `3.0`?.

xster replied:
'Visual effect based on how we see the wheel' is a bit redundant :) 

Everything here visually affects how we see the wheel. 

Try to make the first sentence its own paragraph describing in one line what this does. e.g. 'Distance off-axis from center of the wheel'

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
 
+  /// Visual effect based on how we see the wheel. Imagine we are at some
+  /// distance away from the screen. The wheel is at the center of the screen.
+  /// If we move our position in the xAxis, we will see the wheel differently.
```
'x axis'

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+  /// The centerViewRate is defaulted at 1.0, which is point B. 0.0 will be point A
+  /// and 2.0 will be point C. There's no limit to centerViewRate since we can be
+  /// infinite to the left or right, but a reasonable value would be in the
+  /// range [0.0, 2.0].
```
-infinity <-> 0.0 being center <-> infinity seems like a easier range to work with

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+  double get centerViewRate => _centerViewRate;
+  double _centerViewRate = 1.0;
+  set centerViewRate(double value) {
+    if(value == null) {
```
space after the if

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+  double _centerViewRate = 1.0;
+  set centerViewRate(double value) {
+    if(value == null) {
+      _centerViewRate = 1.0;
```
you still want a check to make sure the value changed from the previously set value and do nothing if the value didn't change

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+      _centerViewRate = 1.0;
+      return;
+    }
+    _centerViewRate = value;
```
you need to either call markNeedsLayout or markNeedsPaint, otherwise this render object wouldn't update.

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+  /// The default value is 1.0, which will not change anything.
+  /// If the value is > 1.0, the center item will be zoomed in by that rate, and
+  /// it will also be rendered as flat, not cylindrical like the rest of the list.
+  double get magnifyRate => _magnifyRate;
```
The word 'rate' is too ambiguous. `magnifyZoom`. 

Ditto above with space, a no change check and markNeedsPaint.

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -243,6 +247,47 @@ class RenderListWheelViewport
+  /// The zoom-in rate of the magnifier.
+  ///
+  /// The default value is 1.0, which will not change anything.
+  /// If the value is > 1.0, the center item will be zoomed in by that rate, and
```
Test and also document what this does when the value is <1.0. 

Also add sanity checks with asserts. What happens when it's -100.0 for instance.

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
-        );
-      },
-    );
+    double magnifierTopLine = size.height / 2 - _itemExtent * _magnifyRate / 2;
```
Let's make this height also configurable like we said yesterday (so it's possible to not have a magnifier at all).

xster replied:
Make the description complete. A line can't be a `double`. Name it magnifierTopLinePosition or something.

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
-      },
-    );
+    double magnifierTopLine = size.height / 2 - _itemExtent * _magnifyRate / 2;
+    double magnifierBotLine = size.height / 2 + _itemExtent * _magnifyRate / 2;
```
Don't abbreviate. Spell out Bottom

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+       && untransformedPaintingCoordinates.dy >= magnifierTopLine -
+                                              _itemExtent*_magnifyRate
+       && untransformedPaintingCoordinates.dy <= magnifierBotLine)
+    {
```
{ on the same line as the closing parenthesis
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+
+    // Some part of the child is in the center magnifier.
+    if(_magnifyRate != 1.0
+       && untransformedPaintingCoordinates.dy >= magnifierTopLine -
```
Line break by priority order of the operations. 
```dart
_magnifyRate != 1.0 
    && untransformedPaintingCoordinates.dy
        >= magnifierTopLine - _itemExtent * _magnifyRate

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+    double magnifierBotLine = size.height / 2 + _itemExtent * _magnifyRate / 2;
+
+    // Some part of the child is in the center magnifier.
+    if(_magnifyRate != 1.0
```
space after if

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+    // Some part of the child is in the center magnifier.
+    if(_magnifyRate != 1.0
+       && untransformedPaintingCoordinates.dy >= magnifierTopLine -
+                                              _itemExtent*_magnifyRate
```
spaces around *

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+      context.pushClipRect(
+        false,
+        offset,
+        new Rect.fromLTWH(0.0, magnifierTopLine,
```
Make this a scoped variable and give it a name so it's easier to understand what this Rect is

xster commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -562,34 +607,104 @@ class RenderListWheelViewport
+        false,
+        offset,
+        new Rect.fromLTWH(0.0, magnifierTopLine,
+                          size.width, _itemExtent * _magnifyRate),
```
When the constructor doesn't fit in one line, make it one parameter per line:
```dart
new Rect.fromLTWH(
  0.0,
  magnifierTopLine,
  size.width,
  _itemExtent * _magnifyRate,
);
```

xster commented on packages/flutter/lib/src/widgets/list_wheel_scroll_view.dart:
```diff
@@ -446,6 +448,12 @@ class ListWheelScrollView extends StatefulWidget {
   /// {@macro flutter.rendering.wheelList.perspective}
   final double perspective;
 
+  /// See[RenderListWheelViewport.centerViewRate]
```
Use `@macro` instead

xster reviewed (commented):
Cool awesome. I did a first pass on the render object.

xster commented:
Instead of git pull-merging, just rebase on head. Since no one else is concurrently working on the same files, it makes it a bit cleaner.

Commit:
cleaning code

Commit:
cleaning code

Commit:
cleaning code

doxuanviet1996 commented on packages/flutter/lib/src/rendering/list_wheel_viewport.dart:
```diff
@@ -247,45 +249,61 @@ class RenderListWheelViewport
-    if(value == null) {
-      _centerViewRate = 1.0;
+  /// {@template flutter.rendering.wheelList.curveRate}
+  /// Affect how much the wheel is curved to the left, or the right.
```
I find it really hard naming and explaining this variable, so if you still feel unclear or have any suggestions, please let me know.

Commit:
fix errors for running test

Commit:
fix errors for running test

Commit:
fix errors in analyze

Commit:
fix errors in analyze

doxuanviet1996 committed:
fix errors in analyze

doxuanviet1996 committed:
Merge branch 'master' of github.com:doxuanviet1996/flutter

doxuanviet1996 committed:
Merge branch 'master' of github.com:doxuanviet1996/flutter

doxuanviet1996 committed:
Merge branch 'master' of github.com:doxuanviet1996/flutter

doxuanviet1996 committed:
Merge branch 'master' of github.com:doxuanviet1996/flutter

Rebase.

doxuanviet1996 committed:
fix analyze errors

doxuanviet1996 committed:
fix analyze errors

doxuanviet1996 committed:
adding multicolumn_picker to gallery

Minor fix in documentation

Add flexes to multicolumn_picker

Fix the magnifier

Clean up code + comments

Minor fix

minor fix

Mark flutter_gallery_instrumentation_test as flaky (#18887)

See #18879.

Fix library-scope expression evaluation. (#18851)

This also improves diagnostic in case of internal compiler errors.

Remove race conditions involving finding available ports (#18698)

This is an attempt to reland #18488 with less breakage on macOS.

Filter invalid characters in the terminal input stream (#18892)

Fixes https://github.com/flutter/flutter/issues/18007

Add --create option to `flutter emulators` command (#18235)

* Add --create option to flutter emulators

* Tweaks to error message

* Simplify emulator search logic

* Make name optional

* Add a note about this option being used with --create

* Tweaks to help information

* Switch to processManager for easier testing

* Don't crash on missing files or missing properties in Android Emulator

* Move name suffixing into emulator manager

This allows it to be tested in the EmulatorManager tests and also used by daemon later if desired.

* Pass the context's android SDK through so it can be mocked by tests

* Misc fixes

* Add tests around emulator creation

Process calls are mocked to avoid needing a real SDK (and to be fast). Full integration tests may be useful, but may require ensuring all build environments etc. are set up correctly.

* Simplify avdManagerPath

Previous changes were to emulatorPath!

* Fix lint errors

* Fix incorrect file exgtension for Windows

* Fix an issue where no system images would crash

reduce throws on an empty collection.

* Fix "null" appearing in error messages

The name we attempted to use will now always be returned, even in the case of failure.

* Add additional info to missing-system-image failure message

On Windows after installing Andriod Studio I didn't have any of these and got this message. Installing with sdkmanager fixed the issue.

* Fix thrown errors

runResult had a toString() but we moved to ProcessResult when switching to ProcessManager to this ended up throwing "Instance of ProcessResult".

* Fix package import

* Fix more package imports

* Move mock implementation into Mock class

There seemed to be issues using Lists in args with Mockito that I couldn't figure out (docs say to use typed() but I couldn't make this compile with these lists still)..

* Rename method that's ambigious now we have create

* Handle where there's no avd path

* Add another toList() :(

* Remove comment that was rewritten

* Fix forbidden import

* Make optional arg more obviously optional

* Reformat doc

* Note that we create a pixel device in help text

* Make this a named arg

Add process output for flutter_tester test and unskip (#18899)

* Add process output for flutter_tester test and unskip

* Add a comment about this potential flake

Add iOS module template (#18830)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Re-skip flutter_device_tester test (#18900)

After landing the un-skip this test failed with a timeout. It then passed on the next build (!). I think it's too flaky to leave running until we can better diagnose what's happening.

Revert "Add iOS module template (#18830)" (#18902)

This reverts commit 8be198d8a086bc26a9b73536f5ebd7ee327a4f01 that broke the build.

Ensure we have the Java binary on PATH when invoking avdmanager (#18908)

See https://github.com/flutter/flutter/issues/13379#issuecomment-400984667.

Add iOS module template (#18830) (#18903)

Add iOS module template

This will enable integration of flutter-views into existing iOS project.

Wait before running the second test. (#18826)

Expose emulator creation to daemon API (#18905)

* Fix typo

* Add emulator.create to Daemon API

* Swap order of daemon changelog entries

Remove the 'app' domain from flutter daemon (#18873)

* Remove the 'app' domain from flutter daemon

By default the daemon won't register the "app" domain, you need to opt-in (which the 'run' command does, as well as the tests for the app functionality).

Fixes #6658.

* Tweak text

* Put restart/callServiceExtension/stop back into daemon mode

* Add a comment about removing discoverApps

Make Flutter module template use FlutterView with less assumptions (#18883)

Remove flaky flutter_tester test (#18916)

This test was designed to ensure flutter_tester keeps running (previously it would quit immediately). However it's turned out ot be rather flaky and we have new tests on the way that supersede this by actually testing real things (debug stepping, reloading, expression evaluation).

Provide workaround for expression compilation frontend issue (#18657)

Add 90th and 99th percentile frame rasterize time for timeline_summary (#18881)

Our first step towards https://github.com/flutter/flutter/issues/18727

Add golden test for text shade overflow (#18816)

Goldens for flutter/flutter#18729

Add 90th and 99th percentiles to dashboard (#18930)

This fixes #18727

Fix typo (#18941)

Set FLUTTER_TEST when spawning flutter_tester (#18907)

Otherwise on Windows these will fail as an unsupported platform when things try to read defaultTargetPlatform in flutter apps.

Retry npm step 5 times with 5 second pause (#18918)

Mark non-flaky tests as such (#18927)

Let the Kernel FE know whether it's building for a release VM. (#18876)

Fixes #18829

Fix intrinsics for ConstrainedBox (#18935)

Fix type errors in iOS simulator discovery. (#18960)

Support all keyboard actions. (#11344) (#18855)

* Support all keyboard actions. (#11344)

Simplify Gradle magic (#18987)

Fix module_test (#18993)

Throw instead of return null if durations is empty (#18957)

See https://github.com/flutter/flutter/pull/18881#discussion_r199285256

[frdb] Change to warning log on timeout. (#19006)

Before a severe log would be raised when timeout happens. Now that
testing connections requires potentially running into a timeout, this
will cause failures when there shouldn't be any.

Show Search Button on keyboard for search widget (#19011)

Also:
 * Export TextInputAction from `text_field.dart` to parallel `TextInputType`.

Nit: add space after for (#18954)

Round properly to nearest page in TabBarView. (#16868)

This logic is described in the test as looking for a scroll ending
very close to a new page, but in fact its behavior is more like
"very close to a page to the right": if we're not very, very close
to any page, it will pick the page to the left, not an old page.

There's no reason this should be left-right asymmetrical.
Instead, pick the nearest page.

In practice, the case where this makes a difference never arises when
the scroll runs undisturbed to completion; but when the user taps on
the page to hold or drag, the scroll will be interrupted before it
gets within tolerance of a particular page, and this case does arise.

This fixes a glitch that is hard to trigger without time dilation,
but is quite conspicuous with it:

 * Open a tab view with at least 4 tabs, e.g. the Buttons screen
   of the gallery (with "Animate Slowly" on.)
 * Starting at tab 0, tap tab 2.
 * When the animation is nearly complete, tap the page a couple
   of times, as if to drag it around to scroll.  Then let the
   page view settle ballistically toward page 2.
 * Before it finishes, tap tab 3.
 * Suddenly page 1 fills the view, replacing page 2, before we
   scroll from there to page 3.

With this fix, the animation in the last step moves smoothly from
where we are when it starts onward to page 3.

Fix type for Dart 2 (#18998)

Fixes #18995.

Include stdout/stderr in failure messages for devicelab's tests (#18990)

* Include stdout/stderr in failure messages

Sometimes some of these tests unexpectedly fail with a non-zero exit code. This ensures stdout/stderr is included in the test failure message when this happens so that we can track down the issue.

* Remove redundant info about exit code

* Remove unnecessary indenting

The [markers] make it fairly clear so this just makes the test code noisy.

flutter attach should handle non-singular device gracefully (#19036)

Upgrade node on travis (#18994)

* Upgrade to current LTS version of Node

The version of NPM with Node 6 is over a year old and frequently hangs trying to install Firebase tools.

* Temporarily comment out this condition so the install runs on Travis

* Revert "Temporarily comment out this condition so the install runs on Travis"

This reverts commit 74db9366b4b1b094369388ef50fb26f3b7fd3104.

* Pin to nodejs v8.1

Add flag to ThemeData to expand tap targets of certain material widgets (#18369)

Make EditableText cursor configurable (#18888)

* fixed segmented control golden test

* fixed segmented control golden test

* added cursorWidth, cursorRadius

* added default value for cursorWidth based on Apple specs

* test default cursorWidth

* removed cursorHeight stuff

* added functionality to keep cursor from blinking

* cursor width and radius is configurable + tests

* changed goldens repo version in goldens.version

* working version of configurable cursor (erased debugKeepCursorOn)

* minor changes

* docs

* changed textfield test that was failing due to new default cursorwidth

* added default value of cursorwidth in RenderEditable

* only run golden file tests on Mac

* cursor tests

* the tests are actually there now

* weak warning fixed

* switching to Linux

* changed default cursorWidth: 2.0 -> 1.0

* assorted changes, including changing text field test

* re-paint -> re-layout when changing cursorWidth

Revert "Add flag to ThemeData to expand tap targets of certain material widgets (#18369)" (#19053)

This reverts commit 989f5741da9e4e8a9ae40d2ec9fbe33f11aba177.

make edit in README.md (#18981)

Fix typos (#19065)

Fix typos (#19067)

Fix typos (#19073)

Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd (#19008)

* Roll engine to 21c7d6a5da47165d076928fbe460badbbead24cd

flutter/engine@21c7d6a Revert "Add antiAlias and saveCount to clipPath and restore (#5638)"
flutter/engine@70dcbb5 Roll src/third_party/skia d818ebf4a317..9aa30c6ee0e5 (11 commits)
flutter/engine@ad42324 Roll src/third_party/skia a219419c9d76..d818ebf4a317 (2 commits)
flutter/engine@f2eb83a Roll src/third_party/skia 00d2e8ebcb13..a219419c9d76 (1 commits)
flutter/engine@0ec7669 Roll src/third_party/skia 8451daabb23d..00d2e8ebcb13 (1 commits)
flutter/engine@061e899 Support all keyboard actions. (#11344)
flutter/engine@bc6b250 Roll src/third_party/skia 34024a7c478c..8451daabb23d (1 commits)
flutter/engine@df4dffb Roll src/third_party/skia 75e69028956d..34024a7c478c (1 commits)
flutter/engine@d8770d4 Roll src/third_party/skia cf863fb9b446..75e69028956d (1 commits)
flutter/engine@b88a8b3 Roll src/third_party/skia b7b9d02ac020..cf863fb9b446 (1 commits)
flutter/engine@2261ccf [fuchsia] Update scenic lib path.
flutter/engine@4c4ef98 [fuchsia] Rename scenic_lib => scenic
flutter/engine@f0c21f3 Roll src/third_party/skia 184d408b646b..b7b9d02ac020 (7 commits)
flutter/engine@a2bf805 Add antiAlias and saveCount to clipPath and restore
flutter/engine@9e450d1 Roll src/third_party/skia c91fe3ab1c5d..184d408b646b (10 commits)
flutter/engine@e663996 Roll src/third_party/skia eb8f8106f38c..c91fe3ab1c5d (5 commits)
flutter/engine@fecd66f Roll src/third_party/skia 723b1f6ef941..eb8f8106f38c (1 commits)
flutter/engine@4466d61 Remove vmservice_io.main from entry points.
flutter/engine@f279dfe Roll src/third_party/skia 7e2327b133db..723b1f6ef941 (1 commits)
flutter/engine@a885bd4 Roll src/third_party/skia 24d18ced1ad7..7e2327b133db (6 commits)
flutter/engine@ad1bd47 Roll src/third_party/skia a1e5630183c1..24d18ced1ad7 (7 commits)
flutter/engine@ce06bba ensure a11y state is communicated back to flutter
flutter/engine@c54a57e Roll src/third_party/skia b1b87d9df81e..a1e5630183c1 (10 commits)
flutter/engine@a9e5354 Roll src/third_party/skia 385804514edf..b1b87d9df81e (6 commits)
flutter/engine@2006e7d Roll src/third_party/skia e97bb26893a1..385804514edf (1 commits)
flutter/engine@2942061 Roll src/third_party/skia e2e52e46ca63..e97bb26893a1 (1 commits)
flutter/engine@e378811 Roll src/third_party/skia 14de25dfc7c4..e2e52e46ca63 (2 commits)
flutter/engine@48cc8f2 Roll src/third_party/skia 551dc3e91143..14de25dfc7c4 (2 commits)
flutter/engine@00ddf13 libtxt: apply an offset to drawing operations instead of translating the canvas
flutter/engine@2c5647c Roll src/third_party/skia 059a9ab4bcd0..551dc3e91143 (5 commits)
flutter/engine@9e44004 Roll src/third_party/skia 45c9dab4c3ec..059a9ab4bcd0 (8 commits)
flutter/engine@fbb3436 Remove some unused code from the Android host
flutter/engine@36f3f95 Roll src/third_party/skia 37b7e4714558..45c9dab4c3ec (1 commits)
flutter/engine@57e2e0c Roll src/third_party/skia c421ca1d6e41..37b7e4714558 (2 commits)
flutter/engine@ce417e2 Roll src/third_party/skia 5b201e3b0ba8..c421ca1d6e41 (3 commits)
flutter/engine@05f5a18 Roll src/third_party/skia d47fe095cf88..5b201e3b0ba8 (6 commits)
flutter/engine@33b70e0 Roll src/third_party/skia 9c0ce41cf711..d47fe095cf88 (12 commits)
flutter/engine@7dab419 Roll src/third_party/skia d4b2adeaa929..9c0ce41cf711 (7 commits)
flutter/engine@d3fa01a Roll src/third_party/skia f46710802ad5..d4b2adeaa929 (1 commits)

Create a CupertinoApp (#18469)

Avoid repeating microbenchmark (#19087)

Attempt to fix race in microbenchmark error handling code (#19092)

Await stream cancellations before trying to complete

Add temporary debug prints to identify devicelab failures (#19093)

Revert "Add temporary debug prints to identify devicelab failures (#19093)" (#19095)

This reverts commit edc16473bfdafc0f5fe546610d3b0fe0eecdd080.

Revert engine roll that caused microbenchmarks failures (#19123)

* Merge

* Update goldens

Adding a setup function for hot restart (#18921)

* Adding a setup function for hot restart that can be overwritten to perform any necessary initialization before a restart is started.

* Adding a test for setupHotRestart

* Removing local pubspec change

* Fixing some type annotation errors

* Responding to comments

Deferred ListView.separator cleanups (#18720)

Fix typo (#19160)

Additional integration test driver functionality (#19085)

* Add additional functionality to integration test driver

* Typo

* Remove delays we shouldn't need + comments

* Fix typos

Fix path of ios module template files (#19037)

* Fix path of ios module template files

* Fix more paths

make tests in xcodeproj_test.dart and gradle_test.dart hermetic (#18093) (#18548)

Re-land #18369 (#19082)

adjust declared types to work with dart 2 typing at runtime (#19007)

* adjust declared types to work with dart 2 typing at runtime

* review comments

* update packages/flutter_tools/lib/src/ios/simulators.dart

remove an unused analysis options file (#19010)

Adjust Cupertino controls's APIs to match Material (#19127)

Put the nav bar padding inside the CustomMultiChildLayout for correct centering and allow custom padding (#19129)

Refer to Navigator.of(rootNavigator) in Cupertino tab widgets (#19130)

Move collectCoverageData into CoverageCollector so it can be re-used. (#19055)

This will let us re-use that code from fuchsia_tester.dart.
Tested by running the stocks example tests with coverage collection before and after.
.lcov files are identical.

Add Windows support to defaultTargetPlatform (#19176)

A significant amonut of code uses defaultTargetPlatform, so it's
currently impossible to run most apps on Windows. This adds a mapping
from Windows as a host to Android as a target, paralleling the
macOS->iOS and Linux->Android mappings.

This allows more use of Windows as a host platform (e.g., for testing,
as in issue #17768).

Add new matcher and utility methods for testing semanics (#19046)

Add getSemanticsId command to flutter_driver (#19047)

Use runTests in fuchsia tester. (#19178)

This allows us to collect code coverage for Fuchsia tests.

Rebuilding BottomNavigationBar when items.length changes (#19179)

Updated the material localizations (#19174)

Add ignoringFeedbackSemantics to Draggable/LongPressDraggable (#18643)

Remove bogus SDK constraint lines. (#19013)

Segmented control animation (#18811)

Added animation to segmented control widget.

Improve flutter tools integration tests (#18865)

* Rename util -> test_utils

* Rename flutter_test_driver -> test_driver

* Switch testWithContext -> test

* Remove unused import

* Move test project into a class to make it easier to have multiple of these

Each "TestProject" class can contain its files and things like named breakpoint locations.

* Split expression evaluation tests into own file

* Include last response in error messages

* Update expectations based on current bugs

* Fix async-ness in tests

* Fix incorrect expectation in test

* Fix incorrect evaluations

* Remove skips for tests that are now passing on master

* Expect pass on Linux

🤷‍♂️

* Call the code

* Skip expression evaluation tests on Windows

* Skip whole group, not just one test

* Remove duplicated method from merge

* Fix misplaced close of group

* Remove code that was duplicated from test we copied

Not sure how this ended up in here?

* Re-fix typo

Improve error message in microbenchmarks when flutter run lives too long (#19200)

This should make it more obvious from the error what's gone wrong if we see a repeat of #19096 (previously the error was something like "Bad state: future already completed").

Fix use of Future<void> as a Future<bool> (#18843)

* Fix use of Future<void> as a Future<bool>

Add segmented control to gallery (#19192)

* Adding segmented control to gallery

doxuanviet1996 committed:
cleaning code

doxuanviet1996 committed:
fix errors for running test

doxuanviet1996 committed:
fix errors in analyze

doxuanviet1996 committed:
.

doxuanviet1996 committed:
Merge branch 'master' of github.com:doxuanviet1996/flutter

doxuanviet1996 committed:
resolve conflict

doxuanviet1996 commented:
Move to #19284
